newapi/ooniapi/measurements.py

```diff
@@ -1335,16 +1294,37 @@ def _clickhouse_aggregation(
     if axis_x == axis_y and axis_x is not None:
         raise ValueError("Axis X and Y cannot be the same")
 
-    def group_by_date(since, until, cols, colnames, group_by):
+    def group_by_date(since, until, time_gran, cols, colnames, group_by):
         if since and until:
-            delta = abs((until - since).days)
+            delta = until - since
         else:
-            delta = 0
+            delta = None
+
+        ranges = (
+            (7, ("hour", "day", "auto")),
+            (30, ("day", "week", "auto")),
+            (365, ("day", "week", "month", "auto")),
+        )
+        if delta is None or delta <= timedelta():
+            raise Exception("Invalid since and until values")
 
-        # TODO: add a granularity parameter and a check around query weight /
-        # / response size. Also add support in CSV format.
-        granularity = ("day", "toDate")
-        tcol, fun = granularity
+        for thresh, allowed in ranges:
+            if delta > timedelta(days=thresh):
+                continue
+            assert time_gran in allowed
```

Comment:
It would be nice in this case to return a meaningful error to the end user stating which are the allowed aggregation intervals for the current time window.

Reply:
Added together with a new set of time ranges.